lib/src/css_mirroring/rulesets_processor.dart

```diff
@@ -21,42 +21,66 @@ import 'css_utils.dart' show Direction;
 import 'entity.dart';
 import 'mirrored_entities.dart';
 import '../utils/enum_parser.dart';
+import 'css_utils.dart'
+    show isDirectionInsensitive, hasNestedRuleSets, Direction, flipDirection;
 
-enum RemovalResult { removedSome, removedAll }
+enum RemovalResult { removedSome, removedAll, removedNone }
 
 /// Returns true if the [RuleSet] was completely removed, false otherwise.
 RemovalResult editFlippedRuleSet(MirroredEntity<RuleSet> mirroredRuleSet,
-    Direction flippedDirection, BufferedTransaction trans) {
-  final subTransaction = trans.createSubTransaction();
+    Direction nativeDirection, BufferedTransaction commonTrans, BufferedTransaction nativeDirTrans, BufferedTransaction flippedDirTrans) {
+  final subTransaction = commonTrans.createSubTransaction();
+  final originalSubTransaction = nativeDirTrans.createSubTransaction();
+  final flippedSubTransaction = flippedDirTrans.createSubTransaction();
 
   MirroredEntities<Declaration> mirroredDeclarations = mirroredRuleSet
       .getChildren((RuleSet r) => r.declarationGroup.declarations);
 
   /// Iterate over Declarations in RuleSet and store start and end points of
   /// declarations to be removed.
-  var removedCount = 0;
+  var commonCount = 0;
+  var flippedCount = 0;
   mirroredDeclarations.forEach((MirroredEntity<Declaration> decl) {
     checkState(decl.flipped.value is Declaration,
         message: () => 'Expected a declaration, got $decl');
 
     if (decl.hasSameTextInBothVersions) {
-      decl.flipped.remove(subTransaction);
-      removedCount++;
+      decl.original.remove(originalSubTransaction);
+      decl.flipped.remove(flippedSubTransaction);
+      commonCount++;
+    } else {
+      decl.original.remove(subTransaction);
+      flippedCount++;
     }
   });
 
-  if (removedCount == mirroredDeclarations.length) {
-    mirroredRuleSet.flipped.remove(trans);
-    return RemovalResult.removedAll;
+  RemovalResult result;
```

Comment:
Consider asserting that flippedCount + commonCount == mirroredDeclarations.length